### PR TITLE
Add NewRelic cross-app transaction tracing

### PR DESCRIPTION
**Pull request by thelollies**
Rory started this one off as a formalised monkey-patch for NewRelic cross-application tracing on AMQP-based inter-resource calls which enabled itself only if NewRelic were present.

We already had some precedent for `LoadError`-based automatic disabling of features based upon optional external components via the `ByConvention` endpoint discovery code and given an emerging pattern along with other backlog items it started to become obvious that we could formalise something grander to allow zero-runtime-impact additions of things like development mode enhanced logging, pretty JSON responses and so-forth. Thus, `Hoodoo::Monkey`.

Taking on Rory's PR, I've refactored it within the monkey pattern. His code becomes the first official reversible monkey patch. There's a drive-by fix of Issue #150 against the `ByConvention` discoverer's use in `Hoodoo:Client` since it was all related to the underlying expected-exception mechanism that started the whole process off in the first place (although that pattern is independent of `Hoodoo:Monkey`'s mechanics).

Note the brand new convention for exposing otherwise-private methods as public monkey patch hook points; simply put `monkey_` on the start of the method name and copy (with appropriate customisation) the comment contents for `monkey_send_request` for the boilerplate stuff referring back to `Hoodoo::Monkey` so that RDoc output remains useful.

### CODE REVIEW

File: lib/hoodoo/client/endpoint/endpoints/amqp.rb [hunk 1]
```diff
@@ -192,7 +193,21 @@ def do_amqp( description_of_request )
             return get_data_for_response( description_of_response )
           end
 
+          # Call Alchemy to with the specified +http_message+. This is extracted
+          # in to its own method to allow it to be wrapped but NewRelic tracing
```
Comment (jeremyolliver) [inline lib/hoodoo/client/endpoint/endpoints/amqp.rb]: Minor typo here but -> by

**Comment by jeremyolliver** [comment]
Code-wise all looks good to me here. the begin/rescue is the best way I can think of to handle automatically loading new relic support while making it optional so :+1: on that too.

**Comment by pond** [comment]
Yep it's all good but *DO NOT MERGE* yet. I'm elevating monkey patching to a first-class citizen (or at least trying to) and this PR will get refactored to incorporate that stuff.

**Comment by pond** [comment]
P.S. To anyone reading - please note that although the code is fine, this wouldn't be permitted for merging as it stands because:

1) It does not have 100% RDoc coverage.
2) There is no RSpec coverage whatsoever.

Patches are only merged if the human code review passes muster, `bundle exec rake rerdoc` still says there's 100% coverage and the RCov coverage report for `bundle exec rspec` reports 100% as well.

**Comment by pond** [comment]
OK, test coverage is back up to 100% and `bundle exec rake rerdoc` reports 100% documentation coverage too, though the RDoc update spam is omitted from this PR. I'll do all that once this PR is eventually merged, as part of a release branch for version 1.8.0.

**Pull request merged**